File: core/lua/astal/widget.lua
local lgi = require("lgi")
local Astal = lgi.require("Astal", "0.1")
local Gtk = lgi.require("Gtk", "3.0")
local GObject = lgi.require("GObject", "2.0")
local Binding = require("astal.binding")
local Variable = require("astal.variable")
local exec_async = require("astal.process").exec_async

local function filter(tbl, fn)
    local copy = {}
    for key, value in pairs(tbl) do
        if fn(value, key) then
            if type(key) == "number" then
                table.insert(copy, value)
            else
                copy[key] = value
            end
        end
    end
    return copy
end

local function map(tbl, fn)
    local copy = {}
    for key, value in pairs(tbl) do
        copy[key] = fn(value)
    end
    return copy
end

local flatten
flatten = function(tbl)
    local copy = {}
    for _, value in pairs(tbl) do
        if type(value) == "table" and getmetatable(value) == nil then
            for _, inner in pairs(flatten(value)) do
                table.insert(copy, inner)
            end
        else
            table.insert(copy, value)
        end
    end
    return copy
end

local function set_children(parent, children)
    children = map(flatten(children), function(item)
        if Gtk.Widget:is_type_of(item) then
            return item
        end
        return Gtk.Label({
            visible = true,
            label = tostring(item),
        })
    end)

    -- remove
    if Gtk.Bin:is_type_of(parent) then
        local rm = parent:get_child()
        if rm ~= nil then
            parent:remove(rm)
        end
    end

    -- FIXME: add rest of the edge cases like Stack
    if Astal.Box:is_type_of(parent) then
        parent:set_children(children)
    elseif Astal.CenterBox:is_type_of(parent) then
        parent.start_widget = children[1]
        parent.center_widget = children[2]
        parent.end_widget = children[3]
    elseif Astal.Overlay:is_type_of(parent) then
        parent:set_child(children[1])
        children[1] = nil
        parent:set_overlays(children)
    elseif Gtk.Container:is_type_of(parent) then
        for _, child in pairs(children) do
            if Gtk.Widget:is_type_of(child) then
                parent:add(child)
            end
        end
    end
end

local function merge_bindings(array)
    local function get_values(...)
        local args = { ... }
        local i = 0
        return map(array, function(value)
            if getmetatable(value) == Binding then
                i = i + 1
                return args[i]
            else
                return value
            end
        end)
    end

    local bindings = filter(array, function(v)
        return getmetatable(v) == Binding
    end)

    if #bindings == 0 then
        return array
    end

    if #bindings == 1 then
        return bindings[1]:as(get_values)
    end

    return Variable.derive(bindings, get_values)()
end

local function astalify(ctor)
    function ctor:hook(object, signalOrCallback, callback)
        if type(object.subscribe) == "function" then
            local unsub = object.subscribe(function(...)
                signalOrCallback(self, ...)
            end)
            self.on_destroy = unsub
            return
        end
        local id = object["on_" .. signalOrCallback](function(_, ...)
            callback(self, ...)
        end)
        self.on_destroy = function()
            GObject.signal_handler_disconnect(object, id)
        end
    end

    function ctor:toggle_class_name(name, on)
        Astal.toggle_class_name(self, name, on)
    end

    return function(tbl)
        if tbl == nil then
            tbl = {}
        end

        local bindings = {}
        local setup = tbl.setup

        -- collect children
        local children = merge_bindings(flatten(filter(tbl, function(_, key)
            return type(key) == "number"
        end)))

        -- default visible to true
        if type(tbl.visible) ~= "boolean" then
            tbl.visible = true
        end

        -- filter props
        local props = filter(tbl, function(_, key)
            return type(key) == "string" and key ~= "setup"
        end)

        -- handle on_ handlers that are strings
        for prop, value in pairs(props) do
            if string.sub(prop, 0, 2) == "on" and type(value) ~= "function" then
                props[prop] = function()
                    exec_async(value, print, print)
                end
            end
        end

        -- handle bindings
        for prop, value in pairs(props) do
            if getmetatable(value) == Binding then
                bindings[prop] = value
                props[prop] = value:get()
            end
        end

        -- construct, attach bindings, add children
        local widget = ctor()

        for prop, value in pairs(props) do
            widget[prop] = value
        end

        for prop, binding in pairs(bindings) do
            widget.on_destroy = binding:subscribe(function(v)
                widget[prop] = v
            end)
        end

        if getmetatable(children) == Binding then
            set_children(widget, children:get())
            widget.on_destroy = children:subscribe(function(v)
                set_children(widget, v)
            end)
        else
            if #children > 0 then
                set_children(widget, children)
            end
        end

        if type(setup) == "function" then
            setup(widget)
        end

        return widget
    end
end

local Widget = {
    astalify = astalify,
    Box = astalify(Astal.Box),
    Button = astalify(Astal.Button),
    CenterBox = astalify(Astal.CenterBox),
    CircularProgress = astalify(Astal.CircularProgress),
    DrawingArea = astalify(Gtk.DrawingArea),
    Entry = astalify(Gtk.Entry),
    EventBox = astalify(Astal.EventBox),
    -- TODO: Fixed
    -- TODO: FlowBox
    Icon = astalify(Astal.Icon),
    Label = astalify(Gtk.Label),
    LevelBar = astalify(Astal.LevelBar),
    -- TODO: ListBox
    Overlay = astalify(Astal.Overlay),
    Revealer = astalify(Gtk.Revealer),
    Scrollable = astalify(Astal.Scrollable),
    Slider = astalify(Astal.Slider),
    -- TODO: Stack
    Switch = astalify(Gtk.Switch),
    Window = astalify(Astal.Window),
}

Gtk.Widget._attribute.css = {
    get = Astal.widget_get_css,
    set = Astal.widget_set_css,
}

Gtk.Widget._attribute.class_name = {
    get = function(self)
        local result = ""
        local strings = Astal.widget_set_class_names(self)
        for i, str in ipairs(strings) do
            result = result .. str
            if i < #strings then
                result = result .. " "
            end
        end
        return result
    end,
    set = function(self, class_name)
        local names = {}
        for word in class_name:gmatch("%S+") do
            table.insert(names, word)
        end
        Astal.widget_set_class_names(self, names)
    end,
}

Gtk.Widget._attribute.cursor = {
    get = Astal.widget_get_cursor,
    set = Astal.widget_set_cursor,
}

Gtk.Widget._attribute.click_through = {
    get = Astal.widget_get_click_through,
    set = Astal.widget_set_click_through,
}

Astal.Box._attribute.children = {
    get = Astal.Box.get_children,
    set = Astal.Box.set_children,
}

return setmetatable(Widget, {
    __call = function(_, ctor)
        return astalify(ctor)
    end,
})

File: core/python/astal/widget.py
import sys
from gi.repository import Astal, Gtk, GObject
from astal.binding import Binding, kebabify


def astalify(klass: type[Gtk.Widget]) -> type:
    def _set_children(self, children):
        if not isinstance(children, list):
            children = [children]

        nchildren = [
            ch if isinstance(ch, Gtk.Widget) else Gtk.Label(visible=True, label=str(ch))
            for ch in children
        ]

        # Remove existing child widgets
        if isinstance(self, Gtk.Bin):
            if ch := self.get_child():
                self.remove(ch)
            if ch is not None and ch not in nchildren:
                ch.destroy()

        elif isinstance(self, Gtk.Container):
            for ch in self.get_children():
                self.remove(ch)
                if ch not in nchildren:
                    ch.destroy()

        # Add children based on the type of container
        if isinstance(self, Astal.Box):
            self.set_children(nchildren)

        elif isinstance(self, Astal.Stack):
            self.set_children(nchildren)

        elif isinstance(self, Astal.CenterBox):
            print("Hello")
            self.start_widget = nchildren[0] if len(nchildren) > 0 else None
            self.center_widget = nchildren[1] if len(nchildren) > 1 else None
            self.end_widget = nchildren[2] if len(nchildren) > 2 else None

        elif isinstance(self, Astal.Overlay):
            child, *overlays = nchildren
            self.set_child(child)
            self.set_overlays(overlays)

        elif isinstance(self, Gtk.Container):
            for ch in nchildren:
                self.add(ch)

    def __init__(self, *_, **kwargs) -> None:
        args = {}
        bindings = {}
        handlers = {}
        children = []
        setup = None

        args["visible"] = kwargs.get("visible", True)

        for key, value in kwargs.items():
            if key == "setup":
                setup = value
            elif isinstance(value, Binding):
                bindings[key] = value
            elif key == "child":
                children = [value]
            elif key == "children":
                children = value
            elif key.startswith("on_"):
                handlers[key] = value
            else:
                args[key] = value

        super(Widget, self).__init__(**args)

        if len(children) > 0:
            _set_children(self, children)

        for key, value in bindings.items():
            setter = (
                lambda v: _set_children(self, v)
                if key == "child" or key == "children"
                else getattr(self, f"set_{key}")
            )

            setter(value.get())
            unsub = value.subscribe(setter)
            self.connect("destroy", lambda _: unsub())

        for key, value in handlers.items():
            sig = kebabify(key.replace("on_", ""))
            if isinstance(value, str):

                def callback(_, task):
                    try:
                        print(Astal.Process.exec_finish(task))
                    except Exception as e:
                        print(e, file=sys.stderr)

                self.connect(sig, lambda *_: Astal.Process.exec_async(value, callback))
            else:
                self.connect(kebabify(key.replace("on_", "")), value)

        if setup:
            setup(self)

    def do_get_property(self, prop):
        if prop.name == "class-name":
            return " ".join(Astal.widget_get_class_names(self))
        elif prop.name == "css":
            return Astal.widget_get_css(self)
        elif prop.name == "cursor":
            return Astal.widget_get_cursor(self)
        elif prop.name == "click-through":
            return Astal.widget_get_click_through(self)
        else:
            return super(Widget, self).do_get_property(prop)

    def do_set_property(self, prop, value):
        if prop.name == "class-name":
            return Astal.widget_set_class_names(self, value.split())
        elif prop.name == "css":
            return Astal.widget_set_css(self, value)
        elif prop.name == "cursor":
            return Astal.widget_set_cursor(self, value)
        elif prop.name == "click-through":
            return Astal.widget_set_click_through(self, value)
        else:
            return super(Widget, self).do_set_property(prop, value)

    Widget = type(
        f"AstalPy{klass.__name__}",
        (klass,),
        {
            "__gtype_name__": f"AstalPy{klass.__name__}",
            "__gproperties__": {
                "class-name": (str, "", "", "", GObject.PARAM_READWRITE),
                "css": (str, "", "", "", GObject.PARAM_READWRITE),
                "cursor": (str, "", "", "default", GObject.PARAM_READWRITE),
                "click-through": (bool, "", "", False, GObject.PARAM_READWRITE),
                "implicit-destroy": (bool, "", "", False, GObject.PARAM_READWRITE),
            },
            "__init__": __init__,
            "do_get_property": do_get_property,
            "do_set_property": do_set_property,
        },
    )

    return Widget


Box = astalify(Astal.Box)
Button = astalify(Astal.Button)
CenterBox = astalify(Astal.CenterBox)
CircularProgress = astalify(Astal.CircularProgress)
DrawingArea = astalify(Gtk.DrawingArea)
Entry = astalify(Gtk.Entry)
EventBox = astalify(Astal.EventBox)
# TODO: Fixed
# TODO: FlowBox
Icon = astalify(Astal.Icon)
Label = astalify(Astal.Label)
LevelBar = astalify(Astal.LevelBar)
# TODO: ListBox
Overlay = astalify(Astal.Overlay)
Revealer = astalify(Gtk.Revealer)
Scrollable = astalify(Astal.Scrollable)
Slider = astalify(Astal.Slider)
Stack = astalify(Astal.Stack)
Switch = astalify(Gtk.Switch)
Window = astalify(Astal.Window)
